### ISSUE TITLE
istio_authn filter missing on gateways

### ISSUE THREAD

**Issue by chaudyg** [istio/istio#21343]
**Bug description**
The istio_authn filter doesn't seem to be applied on envoy gateways, preventing the use of AuthorizationPolicy based on attributes extracted from the client certificate.

**Expected behavior**
The istio_authn filter should be added by default to the gateway listener when using the ISTIO_MUTUAL tls mode.

**Steps to reproduce the bug**
I working on cross cluster mesh, and I am trying to control access at the gateway.

* I enforce STRICT mutual TLS in my cluster
* A Gateway with ISTIO_MUTUAL tls mode
* An AuthorizationPolicy applied to the gateway workload

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: infrabin
  namespace: infrabin
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTPS
    hosts:
    - "*"
    tls:
      mode: ISTIO_MUTUAL
---
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
 name: httpbin
 namespace: istio-system
spec:
 selector:
   matchLabels:
     app: istio-ingressgateway
 rules:
 - from:
   - source:
        namespaces: ['demo']
```

When I inspect the envoy configuration of my gateway pods, the rbac filter is applied to the listener, but the istio_authn plugin isn't. All my requests are denied.

```json
{
     "version_info": "2020-02-21T16:37:48Z/16",
     "listener": {
      "name": "0.0.0.0_80",
      "address": {
       "socket_address": {
        "address": "0.0.0.0",
        "port_value": 80
       }
      },
      "filter_chains": [
       {
        "tls_context": {
         "common_tls_context": {
          "tls_certificates": [
           {
            "certificate_chain": {
             "filename": "/etc/certs/cert-chain.pem"
            },
            "private_key": {
             "filename": "/etc/certs/key.pem"
            }
           }
          ],
          "validation_context": {
           "trusted_ca": {
            "filename": "/etc/certs/root-cert.pem"
           }
          },
          "alpn_protocols": [
           "h2",
           "http/1.1"
          ]
         },
         "require_client_certificate": true
        },
        "filters": [
         {
          "name": "envoy.http_connection_manager",
          "typed_config": {
           "@type": "type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager",
           "stat_prefix": "outbound_0.0.0.0_80",
           "http_filters": [
            {
             "name": "envoy.filters.http.rbac",
             "typed_config": {
              "@type": "type.googleapis.com/envoy.config.filter.http.rbac.v2.RBAC",
              "rules": {
               "policies": {
                "ns[istio-system]-policy[httpbin]-rule[0]": {
                 "permissions": [
                  {
                   "and_rules": {
                    "rules": [
                     {
                      "any": true
                     }
                    ]
                   }
                  }
                 ],
                 "principals": [
                  {
                   "and_ids": {
                    "ids": [
                     {
                      "or_ids": {
                       "ids": [
                        {
                         "metadata": {
                          "filter": "istio_authn",
                          "path": [
                           {
                            "key": "source.principal"
                           }
                          ],
                          "value": {
                           "string_match": {
                            "regex": ".*/ns/demo/.*"
                           }
                          }
                         }
                        }
                       ]
                      }
                     }
                    ]
                   }
                  }
                 ]
                }
               }
              }
             }
            },
```

*current fix*
I managed to solve this issue by adding the following envoyfilter

```yaml
---
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: authn
  namespace: istio-system
spec:
  configPatches:
  - applyTo: HTTP_FILTER
    match:
      context: GATEWAY
      listener:
        portNumber: 80
        filterChain:
          filter:
            name: "envoy.http_connection_manager"
            subFilter:
              name: "envoy.filters.http.rbac"
    patch:
      operation: INSERT_BEFORE
      value:
        name: "istio_authn"
        config:
          "@type": "type.googleapis.com/istio.envoy.config.filter.http.authn.v2alpha1.FilterConfig"
          policy:
            peers:
            - mtls: {}
```
This is the result:
```json
{
     "version_info": "2020-02-21T16:42:56Z/17",
     "listener": {
      "name": "0.0.0.0_80",
      "address": {
       "socket_address": {
        "address": "0.0.0.0",
        "port_value": 80
       }
      },
      "filter_chains": [
       {
        "tls_context": {
         "common_tls_context": {
          "tls_certificates": [
           {
            "certificate_chain": {
             "filename": "/etc/certs/cert-chain.pem"
            },
            "private_key": {
             "filename": "/etc/certs/key.pem"
            }
           }
          ],
          "validation_context": {
           "trusted_ca": {
            "filename": "/etc/certs/root-cert.pem"
           }
          },
          "alpn_protocols": [
           "h2",
           "http/1.1"
          ]
         },
         "require_client_certificate": true
        },
        "filters": [
         {
          "name": "envoy.http_connection_manager",
          "typed_config": {
           "@type": "type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager",
           "stat_prefix": "outbound_0.0.0.0_80",
           "http_filters": [
            {
             "name": "istio_authn",
             "config": {
              "@type": "type.googleapis.com/istio.envoy.config.filter.http.authn.v2alpha1.FilterConfig",
              "policy": {
               "peers": [
                {
                 "mtls": {}
                }
               ]
              }
             }
            },
            {
             "name": "envoy.filters.http.rbac",
             "typed_config": {
              "@type": "type.googleapis.com/envoy.config.filter.http.rbac.v2.RBAC",
              "rules": {
               "policies": {
                "ns[istio-system]-policy[httpbin]-rule[0]": {
                 "permissions": [
                  {
                   "and_rules": {
                    "rules": [
                     {
                      "any": true
                     }
                    ]
                   }
                  }
                 ],
                 "principals": [
                  {
                   "and_ids": {
                    "ids": [
                     {
                      "or_ids": {
                       "ids": [
                        {
                         "metadata": {
                          "filter": "istio_authn",
                          "path": [
                           {
                            "key": "source.principal"
                           }
                          ],
                          "value": {
                           "string_match": {
                            "regex": ".*/ns/demo/.*"
                           }
                          }
                         }
                        }
                       ]
                      }
                     }
                    ]
                   }
                  }
                 ]
                }
               }
              }
             }
            },
...
```

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

```
client version: 1.4.2
control plane version: 1.4.3
data plane version: 1.4.3 (2 proxies)
```

**How was Istio installed?**

Official helm chart.
```yaml
# Istio chart values
global:
  controlPlaneSecurityEnabled: true
  mtls:
    enabled: true
  proxy:
    accessLogFile: /dev/stdout
    accessLogEncoding: JSON
mixer:
  policy:
    enabled: true
  telemetry:
    enabled: true

```
**Environment where bug was observed (cloud vendor, OS, etc)**
local kind cluster `v1.17.0`

**Comment by yangminzhu** [comment]
The authn filter is currently not supported on gateway and this is a known limitation for using authz on gateway. The workaround is to use the EnvoyFilter just as you described.

I think the short term solution is probably to investigate is it safe to just add the authn filter on the gateway. The better long term solution is to further decouple the dependency on the authn filter so that the authZ policy could be used independently.

cc @diemtvu

**Comment by diemtvu** [comment]
Actually, we do support this in 1.5, as the new policy is now on workload. I've just finished the tutorial exactly for this usecase. Please check the [prelim doc](https://preliminary.istio.io/docs/tasks/security/authentication/authn-policy/#end-user-authentication) and let me know if you see any problem. Thanks.

**Comment by chaudyg** [comment]
Thx @diemtvu . I will give it a try with the 1.5 release.

I did manage to apply the AuthorizationPolicy on the gateway in 1.4. As you can see above, the RBAC filter has been applied to the envoy configuration. But the auth_n filter is missing, and attributes are not extracted from the client certs).

The tutorial looks good. Thx for that. Are the JWT authentication attributes also extracted using the authn filter behind the scene?

**Comment by incfly** [comment]
They are the same implementation as the workload, 1.5 is out, please give it a try. I'm closing this one. Feel free to create new one against 1.5 support if you identify some issue.
thx.

**Comment by chaudyg** [comment]
@incfly Sorry for the delay (covid-19 with kids doesn't help)

The issue is still there is in istio 1.5.1.

```
$ istioctl version --remote
client version: 1.5.1
control plane version: 1.5.1
data plane version: 1.5.1 (2 proxies)
```

```yaml
{
     "name": "0.0.0.0_80",
     "active_state": {
      "version_info": "2020-04-04T10:30:00Z/6",
      "listener": {
       "@type": "type.googleapis.com/envoy.api.v2.Listener",
       "name": "0.0.0.0_80",
       "address": {
        "socket_address": {
         "address": "0.0.0.0",
         "port_value": 80
        }
       },
       "filter_chains": [
        {
         "filters": [
          {
           "name": "envoy.http_connection_manager",
           "typed_config": {
            "@type": "type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager",
            "stat_prefix": "outbound_0.0.0.0_80",
            "rds": {
             "config_source": {
              "ads": {}
             },
             "route_config_name": "https.80.http.infrabin.infrabin"
            },
            "http_filters": [
             {
              "name": "envoy.filters.http.rbac",
              "typed_config": {
               "@type": "type.googleapis.com/envoy.config.filter.http.rbac.v2.RBAC",
               "rules": {
                "policies": {
                 "ns[istio-system]-policy[httpbin]-rule[0]": {
                  "permissions": [
                   {
                    "and_rules": {
                     "rules": [
                      {
                       "any": true
                      }
                     ]
                    }
                   }
                  ],
                  "principals": [
                   {
                    "and_ids": {
                     "ids": [
                      {
                       "or_ids": {
                        "ids": [
                         {
                          "metadata": {
                           "filter": "istio_authn",
                           "path": [
                            {
                             "key": "source.principal"
                            }
                           ],
                           "value": {
                            "string_match": {
                             "safe_regex": {
                              "google_re2": {},
                              "regex": ".*/ns/demo/.*"
                             }
                            }
                           }
                          }
                         }
                        ]
                       }
                      }
                     ]
                    }
                   }
                  ]
                 }
                }
               }
              }
             },
...
```

As you can see, the  the `rbac` filter is applied to the listener on the gateway pod, but the `istio_authn` plugin isn't. Authentication based on x509 certificates does not work on the gateway, and the `rbac` rules do not apply.   

Could we re-open this issue please? I will try to find some time to dig into the codebase as well.

Many thanks, georges

**Comment by yangminzhu** [comment]
> Actually, we do support this in 1.5, as the new policy is now on workload. I've just finished the tutorial exactly for this usecase. Please check the [prelim doc](https://preliminary.istio.io/docs/tasks/security/authentication/authn-policy/#end-user-authentication) and let me know if you see any problem. Thanks.

@diemtvu I believe @chaudyg is actually trying to use PeerAuthentication (not RequestAuthentication) on Ingress gateway and I think that is not supported, could you confirm? Thanks.

**Comment by diemtvu** [comment]
I see. Peer authn (aka Istio mTLS) is not designed for gateway, thus it doesn't add authn filter as expected.
On the other hand, you can set TLS/mTLS via gateway API. This however doesn't add the authn filter neither. I agree that this is confusing. Feel free to create a feature request issue. I think we might revisit this soon while re-evaluating authn filter WASM.

**Comment by incfly** [comment]
> I see. Peer authn (aka Istio mTLS) is not designed for gateway, thus it doesn't add authn filter as expected.

Could you remind a bit of impl details bit here? Is this just a coding/impl issue? I mean because we only install PeerAuthn on sidecar LDS configuration path, not for the gateway. Thus it does not show up there? To me, adding this support does not seem complicated, but maybe more of question of whether we should add. is that right?

On the other hand, @chaudyg might want to have the benefits of `authn_filter` attribute transformation, such that the those attributes `request.x` can be used for authorization. Gateway mTLS does not help with that.

I think maybe you can look into `EnvoyFilter` to explicitly add that to gateway, giving time being's limitation? @chaudyg

**Comment by chaudyg** [comment]
> On the other hand, you can set TLS/mTLS via gateway API. This however doesn't add the authn filter neither. I agree that this is confusing. Feel free to create a feature request issue. I think we might revisit this soon while re-evaluating authn filter WASM.

Confusing indeed.  I will raise a  feature request issue.

> On the other hand, @chaudyg might want to have the benefits of authn_filter attribute transformation, such that the those attributes request.x can be used for authorization. Gateway mTLS does not help with that.

Correct. For context, my goal is to inter-connect services running on different clusters via a secure gateway.

> I think maybe you can look into EnvoyFilter to explicitly add that to gateway, giving time being's limitation? @chaudyg

The following `EnvoyFilter` (mentioned above) works for now. I guess the real question is whether or not this will eventually be supported and tested. I don't want to box myself into a corner.

```
---
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: authn
  namespace: istio-system
spec:
  configPatches:
  - applyTo: HTTP_FILTER
    match:
      context: GATEWAY
      listener:
        portNumber: 80
        filterChain:
          filter:
            name: "envoy.http_connection_manager"
            subFilter:
              name: "envoy.filters.http.rbac"
    patch:
      operation: INSERT_BEFORE
      value:
        name: "istio_authn"
        config:
          "@type": "type.googleapis.com/istio.envoy.config.filter.http.authn.v2alpha1.FilterConfig"
          policy:
            peers:
            - mtls: {}
```

**Comment by chaudyg** [comment]
Created the feature request. See https://github.com/istio/istio/issues/23284

**Comment by harishkashyap569** [comment]
I'm facing the same issue. i'm using istio 1.4.7 and `istio_authn` missing the claims object in it. 
filter_metadata {
  key: "istio_authn"
  value {
    fields {
      key: "request.auth.principal"
      value {
        string_value: "dfac-sre-tools-npe-981e.svc.id.goog/ns/istio-system/sa/istio-ingressgateway-service-account"
      }
    }
    fields {
      key: "source.namespace"
      value {
        string_value: "istio-system"
      }
    }
    fields {
      key: "source.principal"
      value {
        string_value: "dfac-sre-tools-npe-981e.svc.id.goog/ns/istio-system/sa/istio-ingressgateway-service-account"
      }
    }
    fields {
      key: "source.user"
      value {
        string_value: "dfac-sre-tools-npe-981e.svc.id.goog/ns/istio-system/sa/istio-ingressgateway-service-account"
      }
    }
  }
}

**Comment by howardjohn** [comment]
This has been fixed by #25818 cc @rshriram
